realtek: add support for DGS-1210-52

luizluca opened the pull request:
Hardware specification
----------------------

* RTL8393M SoC, 1 MIPS 34Kc core @ 700MHz
* 128MB DRAM
* 32MB NOR Flash
* 48 x 10/100/1000BASE-T ports
  - 6 x External PHY with 8 ports (RTL8218D)
* 4 x Gigabit RJ45/SFP Combo ports
- External PHY with 4 SFP ports (RTL8214FC) (accoding to u-boot)
* Power LED
* Reset button on front panel
* UART (115200 8N1) via unpopulated standard 0.1" pin header marked J14

The reset-gpio is not required but it does reset the switch.

The 4 combo ports were not stable and were commented out. The RTL8214FC
is what the u-boot reported. However, the Linux kernel reported 49 and
50 as "External RTL8393 SERDES" and 51 and 52 as "RTL8218B (external)".
Those ports only work if u-boot initialize it (for example, loading
initramfs image using one of those ports).

The firmware recovery using U-Boot is broken for all DGS-1210 tested
devices as pressing RESET does not trigger it (only if pressed from a
running stock image)

UART pinout
-----------

[o]ooo|J14
 | ||`------ GND
 | |`------- RX
 | `-------- TX
 `---------- Vcc (3V3)

Installation using OEM upgrade
------------------------------

1. Make sure you are running OEM firmware in image2 slot (logged as admin):
   - > config firmware image_id 2 boot_up
   - > reboot
2. Install squashfs-factory_image1.bin to image1 using (logged as admin):
   - > download firmware_fromTFTP <tftpserver> factory_image1.bin
   - > config firmware image_id 1 boot_up
   - > reboot

Installation using serial interface
-----------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Init network with `rtk network on` command
4. Load image with `tftpboot 0x8f000000 openwrt-realtek-rtl838x-d-link_dgs-1210-52-initramfs-kernel.bin` command
5. Boot the image with `bootm` command

Once booted the initramfs, install the squashfs-sysupgrade.bin as a
normal OpenWrt system.

Dual-boot with stock firmware using writable u-boot-env
-------------------------------------------------------

From stock to OpenWrt / boot image 1 (CLI as admin):
   - > config firmware image_id 1 boot_up
   - > reboot

From OpenWrt to stock / boot image 2: (shell as root)
   - # fw_setenv bootcmd 'run addargs ; bootm 0xb4e80000'
   - # fw_setenv image '/dev/mtdblock7'
   - # reboot

Debrick using serial interface
------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

Back to stock firmware using dual-boot
--------------------------------------

If you have serial interface, you can change u-boot env vars
interrupting the boot process. If not but you are running OpenWrt, you
can dual-boot (as mentioned eariler) and skip to step 4:

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Boot the image 2:
   - set image /dev/mtdblock7; run addargs; bootm 0xb4e80000
4. Once booted, log as admin and change the boot image to 2
   - > config firmware image_id 2 boot_up
   - > reboot
5. After the boot, flash image1 with the vendor image

Back to stock firmware using DNA
--------------------------------

1. From an OpenWrt:
   - # fw_setenv bootstop on
   - # reboot
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

It has been developed and tested on device with F3 revision.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

luizluca committed:
realtek: add support for DGS-1210-52

Hardware specification
----------------------

* RTL8393M SoC, 1 MIPS 34Kc core @ 700MHz
* 128MB DRAM
* 32MB NOR Flash
* 48 x 10/100/1000BASE-T ports
  - 6 x External PHY with 8 ports (RTL8218D)
* 4 x Gigabit RJ45/SFP Combo ports
  - External PHY with 4 SFP ports (RTL8214FC) (accoding to u-boot)
* Power LED
* Reset button on front panel
* UART (115200 8N1) via unpopulated standard 0.1" pin header marked J14

The gpio-restart node is not required but it does reset the switch.

The 4 combo ports were not stable and were commented out. The RTL8214FC
is what the u-boot reported. However, the Linux kernel reported 49 and
50 as "External RTL8393 SERDES" and 51 and 52 as "RTL8218B (external)".
Those ports only work if u-boot initialize it (for example, loading
initramfs image using one of those ports).

The firmware recovery using U-Boot is broken for all DGS-1210 tested
devices as pressing RESET does not trigger it (only if pressed from a
running stock image)

UART pinout
-----------

[o]ooo|J14
 | ||`------ GND
 | |`------- RX
 | `-------- TX
 `---------- Vcc (3V3)

Installation using OEM upgrade
------------------------------

1. Make sure you are running OEM firmware in image2 slot (logged as admin):
   - > config firmware image_id 2 boot_up
   - > reboot
2. Install squashfs-factory_image1.bin to image1 using (logged as admin):
   - > download firmware_fromTFTP <tftpserver> factory_image1.bin
   - > config firmware image_id 1 boot_up
   - > reboot

Installation using serial interface
-----------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Init network with `rtk network on` command
4. Load image with `tftpboot 0x8f000000 openwrt-realtek-rtl839x-d-link_dgs-1210-52-initramfs-kernel.bin` command
5. Boot the image with `bootm` command

Once booted the initramfs, install the squashfs-sysupgrade.bin as a
normal OpenWrt system.

Dual-boot with stock firmware using writable u-boot-env
-------------------------------------------------------

From stock to OpenWrt / boot image 1 (CLI as admin):
   - > config firmware image_id 1 boot_up
   - > reboot

From OpenWrt to stock / boot image 2: (shell as root)
   - # fw_setenv bootcmd 'run addargs ; bootm 0xb4e80000'
   - # fw_setenv image '/dev/mtdblock7'
   - # reboot

Debrick using serial interface
------------------------------

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

Back to stock firmware using dual-boot
--------------------------------------

If you have serial interface, you can change u-boot env vars
interrupting the boot process. If not but you are running OpenWrt, you
can dual-boot (as mentioned eariler) and skip to step 4:

1. Press Escape key during `Hit Esc key to stop autoboot` prompt
2. Press CTRL+C keys to get into real U-Boot prompt
3. Boot the image 2:
   - set image /dev/mtdblock7; run addargs; bootm 0xb4e80000
4. Once booted, log as admin and change the boot image to 2
   - > config firmware image_id 2 boot_up
   - > reboot
5. After the boot, flash image1 with the vendor image

Back to stock firmware using DNA
--------------------------------

1. From an OpenWrt:
   - # fw_setenv bootstop on
   - # reboot
2. In a Windows PC, run 'D-Link Network Assistant v2.0.2.4'. It should
   detect the switch
3. Flash the firmware.

It has been developed and tested on device with F3 revision.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

svanheule commented on target/linux/realtek/image/rtl839x.mk:
```diff
@@ -1,5 +1,26 @@
+  KERNEL := kernel-bin | append-dtb | gzip | uImage gzip | dlink-cameo
+  IMAGES += factory_image1.bin
+  IMAGE/factory_image1.bin := append-kernel | pad-to 64k | \
+        append-rootfs | pad-rootfs | pad-to 16 | check-size | \
```
Usually indented with one tab

svanheule commented on target/linux/realtek/image/rtl839x.mk:
```diff
@@ -1,5 +1,26 @@
+  DEVICE_VENDOR := D-Link
+  FACTORY_IMAGE_SIZE := 12774720
+  DEVICE_MODEL := DGS-1210-52
+  DLINK_KERNEL_PART_SIZE := 1572864
```
Since most (or all) of the factory image parameter are shared with the rtl838x images, could you mode this to a Device/d-link_dgs-1210 directive in the main Makefile for the target? The `SOC :=` assignments will have to be moved to the specific devices though.

luizluca replied:
OK. I might have converted it into spaces during copy/paste.

svanheule replied:
I don't think it's worth the hassle to define that for just the shared `SOC`, so no d-link_dgs1210_rtl8382 needed.

svanheule replied:
IMHO a default should only be added if it makes sense for all devices. Defining a `SOC` value that must be overwritten is just waiting for someone to forget, and possibly being confused by the issues that causes.

luizluca replied:
OK. That's what I did. Thanks!

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,371 @@
+		label = SWITCH_PORT_LABEL(s) ; \
+		phy-handle = <&phy##n>; \
+		phy-mode = #m ; \
+	};
```
Perhaps these macros should be moved to a common include for the realtek target, so they don't have to be defined for every device/subtarget.

luizluca replied:
I don't know if I should create them based on a device where it does not work (I believe because the driver requires adaptations). Once DGS-1210-16 SFP ports support is submitted and merged, it would be more legit to add that new macro file for others to import, incluiding this one.

I just included those macros to keep the SFP information I have already gathered anchored to a valid DTS file.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,371 @@
+
+	sda-gpios, scl-gpios, los-gpio and other settings are just guesses based on dsg-1210-16
+	from https://github.com/openwrt/openwrt/commit/b05cc25704f554d73b8b6772e844c3afddc4821c
+	*/
```
Could you stick to the kernel style for block comments?

```C
/*
 * Every line is prefixed with '*' so it's more clear where the block ends
 */
```

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,371 @@
+	from https://github.com/openwrt/openwrt/commit/b05cc25704f554d73b8b6772e844c3afddc4821c
+	*/
+
+	// Lan 49
```
Below (old) C-style comments are used, please use them here too.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,371 @@
+		i2c-bus = <&i2c0>;
+		los-gpio = <&gpio1 9 GPIO_ACTIVE_HIGH>;
+		mod-def0-gpio = <&gpio1 8 GPIO_ACTIVE_LOW>;
+		// tx-disable-gpio handled by RTL8214FC based on media setting
```
... and here. Basically for all the comments below.

svanheule commented on target/linux/realtek/dts-5.10/rtl8393_d-link_dgs-1210-52.dts:
```diff
@@ -0,0 +1,371 @@
+
+&gpio0 {
+	indirect-access-bus-id = <0>;
+};
```
You can drop this, it doesn't do anything any more. The RTL8231 is already defined by your `gpio1` node.

svanheule reviewed (changes requested):
`- External PHY with 4 SFP ports (RTL8214FC) (accoding to u-boot)`
Missing indentation

`The reset-gpio is not required but it does reset the switch.`
"gpio-restart node"

`4. Load image with "tftpboot 0x8f000000 openwrt-realtek-rtl838x-d-link_dgs-1210-52-initramfs-kernel.bin" command`
Image name will contain `rtl839x`, not `rtl838x`.

Mostly minor things really. Moving the preprocessor macros and "Device/d-link_dgs-1210" block should probably go into separate commits.

luizluca committed:
realtek: move common DGS-1210 def to Makefile

The same parameters for DGS-1210 based on rtl838x SoC are shared with
models with rtl839x SoC.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
BTW, I randomly get this error during boot. It is a little bit rare but it happens:

```
[    1.583470] irq 0: nobody cared (try booting with the "irqpoll" option)
[    1.590875] CPU: 0 PID: 1033 Comm: kworker/u4:8 Not tainted 5.10.127 #0
[    1.598250] Stack : 806aaf4c 80725df8 80740000 800805e4 807b0000 806abc84 00000000 00000000
[    1.607638]         8200bdcc 808f0000 80684284 829a3898 80734be7 00000001 8200bd70 d5f33898
[    1.617028]         00000000 00000000 80684284 8200bc08 ffffefff 00000000 00000000 ffffffea
[    1.626408]         0000006b 8200bc14 0000006b 8073a128 00000000 00000000 00000000 80680000
[    1.635790]         80724d0c 00000008 807a96f4 80725df8 00000018 8035de4c 00000000 808f0000
[    1.645178]         ...
[    1.647920] Call Trace:
[    1.650664] [<80006d68>] show_stack+0x30/0x100
[    1.655638] [<80303524>] dump_stack+0x9c/0xcc
[    1.660513] [<80086e8c>] __report_bad_irq+0x5c/0xf4
[    1.665974] [<800873e0>] note_interrupt+0x2cc/0x31c
[    1.671427] [<80083ae4>] handle_irq_event_percpu+0x5c/0x74
[    1.677567] [<8008938c>] handle_percpu_irq+0x88/0xb8
[    1.683124] [<80082fe8>] generic_handle_irq+0x44/0x5c
[    1.688781] [<80335430>] realtek_gpio_irq_handler+0xc8/0x1a8
[    1.695110] [<80082fe8>] generic_handle_irq+0x44/0x5c
[    1.700767] [<8031f634>] realtek_irq_dispatch+0x94/0x158
[    1.706708] [<80082fe8>] generic_handle_irq+0x44/0x5c
[    1.712379] [<805e3ab4>] do_IRQ+0x1c/0x2c
[    1.716862] [<8031ed3c>] plat_irq_dispatch+0x68/0xf0
[    1.722418] [<800020a8>] except_vec_vi_end+0xb8/0xc4
[    1.727981] [<800b8060>] smp_call_function_single+0x158/0x1f0
[    1.734414] [<800b8700>] smp_call_function+0x24/0x30
[    1.739964] [<80010284>] flush_tlb_mm+0x4c/0xec
[    1.745035] [<8016ca84>] tlb_finish_mmu+0x16c/0x1bc
[    1.750485] [<8016a4cc>] exit_mmap+0x100/0x1ec
[    1.755470] [<80029bc8>] mmput+0x48/0xfc
[    1.759868] [<801abe04>] free_bprm+0x28/0xc0
[    1.764651] [<801ae0f8>] kernel_execve+0x150/0x1c8
[    1.770014] [<80043e98>] call_usermodehelper_exec_async+0x110/0x190
[    1.777022] [<80001b98>] ret_from_kernel_thread+0x14/0x1c
[    1.783060] 
[    1.784724] handlers:
[    1.787250] Disabling IRQ #0
```

It loops over and over this. Does anyone have a clue?

svanheule commented:
> `[ 1.583470] irq 0: nobody cared (try booting with the "irqpoll" option)`

IIRC IRQ 0 is not actually a valid IRQ number under Linux. I'm not sure what could be causing this, so I'll need to have a look at our interrupt handler. I suspect it's due to some spurious interrupts, but could also be incomplete setup somewhere.

luizluca commented:
I'm converting into a draft. The vlan filtering is not working at all. All ports are receiving all vlans, with untagged vlans converging into a single shared broadcast domain.

svanheule commented:
I've noticed this VLAN issue too. Was trying to test some other patches (https://patchwork.ozlabs.org/user/todo/openwrt/?series=305445) last week, but I couldn't get the VLAN tags to show up at all. Same behaviour with and without those patches...

luizluca commented:
> I've noticed this VLAN issue too. Was trying to test some other patches (https://patchwork.ozlabs.org/user/todo/openwrt/?series=305445) last week, but I couldn't get the VLAN tags to show up at all. Same behaviour with and without those patches...

I noticed some extra dmesg info that I didn't see with other devices:

```
[  957.563103] -> index 5116, key 4ff, bucket 0, dmac 0000647002cc2d78, fid: 3c rvid: c
[  957.571851] MAC: 64:70:02:cc:2d:78 vid: 12, rvid: 12, port: 47, valid: 1
[  957.579841] Type: 1, is_static: 0, is_ip_mc: 0, is_ipv6_mc: 0, block_da: 0
[  957.588026]   block_sa: 0, susp: 0, nh: 0, age: 3, is_trunk: 0, trunk: 0
[  957.595936]   stack_dev: 227
[  957.599676] seed: 000c647002cc2d78, key based on rvid: 000004ff
[  957.611534] -> index 9208, key 8fe, bucket 0, dmac 0000647002cc2d78, fid: 7c rvid: 7c
[  957.620330] MAC: 64:70:02:cc:2d:78 vid: 0, rvid: 124, port: 47, valid: 1
[  957.627888] Type: 1, is_static: 0, is_ip_mc: 0, is_ipv6_mc: 0, block_da: 0
[  957.635998]   block_sa: 0, susp: 0, nh: 0, age: 3, is_trunk: 0, trunk: 0
[  957.643870]   stack_dev: 227
[  957.647729] seed: 007c647002cc2d78, key based on rvid: 000008fe
[  957.661795] -> index 15344, key efc, bucket 0, dmac 0000647002cc2d78, fid: f4 rvid: e4
[  957.670684] MAC: 64:70:02:cc:2d:78 vid: 228, rvid: 228, port: 47, valid: 1
[  957.678870] Type: 1, is_static: 0, is_ip_mc: 0, is_ipv6_mc: 0, block_da: 0
[  957.686951]   block_sa: 0, susp: 0, nh: 0, age: 2, is_trunk: 0, trunk: 0
[  957.694517]   stack_dev: 227
[  957.698263] seed: 00e4647002cc2d78, key based on rvid: 00000efc
```

Maybe it is related. I do see vlan tags but the switch is failing to learn MACs. Any port is receiving all packets for all vlans, even with just some vlans configured. I have:

- lan1: untagged (vlan124), tagged (vlan228)
- lan2: untagged (vlan124), tagged (vlan228)
- lan47: untagged (vlan12), tagged (vlan669)
- lan48: untagged (vlan124), tagged (vlan12, vlan228, vlan669)

The switch is mixing vlan124 and vlan12 broadcast domains. And I get tagged vlan12 in lan1 and lan2 (and probably also other vlans that not in use). It's messed up.

```
# for lan in lan1 lan2 lan47 lan48; do echo $lan; bridge fdb show dev $lan; done
lan1
a4:2a:95:1d:26:30 master switch permanent
f8:1a:67:00:16:28 self 
lan2
a6:2a:95:1d:26:31 vlan 124 master switch permanent
a6:2a:95:1d:26:31 vlan 228 master switch permanent
a6:2a:95:1d:26:31 master switch permanent
lan47
a6:2a:95:1d:26:5e vlan 669 master switch permanent
a6:2a:95:1d:26:5e vlan 12 master switch permanent
a6:2a:95:1d:26:5e master switch permanent
lan48
a6:2a:95:1d:26:5f vlan 124 master switch permanent
64:70:02:cc:2d:78 vlan 12 master switch 
a6:2a:95:1d:26:5f vlan 669 master switch permanent
a6:2a:95:1d:26:5f vlan 228 master switch permanent
a6:2a:95:1d:26:5f vlan 12 master switch permanent
a6:2a:95:1d:26:5f master switch permanent
64:70:02:cc:2d:78 vlan 669 self 
64:70:02:cc:2d:78 vlan 12 self 
64:70:02:cc:2d:78 self 
64:70:02:cc:2d:78 vlan 228 self 
```

From that, vlans are missing in lan1 (although the conf is identical to lan2). But it does not really matter as I get all vlans in all ports.

> It's up to you if you want to postpone merging this device until the VLAN issue is fixed, but as far as I'm concerned that's a different problem.

Is it valid to merge even with broken address learning?

svanheule commented:
The patches I linked earlier are supposed to fix MAC learning and VLAN tagging on the first port. I suppose you performed these test without those patches? If so, could you also apply my patches and let me know if you see any difference (hopefully improvement) in behaviour?

svanheule commented:
Apparently I posted a private link, but the public link to the patch series is not very different:
https://patchwork.ozlabs.org/project/openwrt/list/?series=305445

luizluca commented:
> Apparently I posted a private link, but the public link to the patch series is not very different: https://patchwork.ozlabs.org/project/openwrt/list/?series=305445

Those patches fixed most issues! Port 1 vlan is now listed and I get only the traffic for my vlans, including tagged. The issue is that those packets that should be tagged are not tagged.

I have some experience with rtl8367c switches (linux upstream) and I could play a little bit to try to help with this switch. Is there any public programming guide for that SoC? I'll be away for 2 weeks but I can take a look after that.

BTW, is there any progress to get these switches upstream? It looks like it will take a lot of work and cleanup. Tag is done in the ethernet driver while upstream implements dedicated taggers for each protocol. And the switch will probably need to use regmap.

svanheule commented:
> Those patches fixed most issues! Port 1 vlan is now listed and I get only the traffic for my vlans, including tagged. The issue is that those packets that should be tagged are not tagged.
> 
>  have some experience with rtl8367c switches (linux upstream) and I could play a little bit to try to help with this switch. Is there any public programming guide for that SoC? I'll be away for 2 weeks but I can take a look after that.

Good to know the patches actually help. Thanks for testing! Let me know if I can add a Tested-by.

Realtek sadly enough has a tendency of giving products very similar names, and very different implementations. The register layouts for rtl838x/rtl839x/rtl930x/rtl931x are completely different. I've created a website around the switchcore register definition tables extracted from the SDKs: https://svanheule.net/realtek/
The code archives this info is drawn from, and on which the current implementation is based, can be found here: https://svanheule.net/switches/gpl_source_drops

And that's about it, as far as documentation goes. On the register website, we fill in what we can figure out from the code or hardware behaviour. If you would like to help out with that effort, I can add your Github account to the user whitelist.

> BTW, is there any progress to get these switches upstream? It looks like it will take a lot of work and cleanup. Tag is done in the ethernet driver while upstream implements dedicated taggers for each protocol.

There was some initial work on SPI, and IRQ peripherals. I have mainly been working on GPIO (internal and RTL8231). Work on the ethernet part is done entirely downstream by others, with more focus on getting things up and running. I know about some issues, like the tagging, but there will surely be plenty of others.

> And the switch will probably need to use regmap.

I am aware and started to work on some of the smaller parts of the switchcore. See https://github.com/svanheule/linux/tree/leds/rtl-switch-port which I hope to provide as an RFC to OpenWrt this weekend.

luizluca commented:
> > Those patches fixed most issues! Port 1 vlan is now listed and I get only the traffic for my vlans, including tagged. The issue is that those packets that should be tagged are not tagged.
> > 
> >  have some experience with rtl8367c switches (linux upstream) and I could play a little bit to try to help with this switch. Is there any public programming guide for that SoC? I'll be away for 2 weeks but I can take a look after that.
> 
> Good to know the patches actually help. Thanks for testing! Let me know if I can add a Tested-by.

Besides the untagged packets that should be tagged:

Tested-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

> 
> Realtek sadly enough has a tendency of giving products very similar names, and very different implementations. The register layouts for rtl838x/rtl839x/rtl930x/rtl931x are completely different. I've created a website around the switchcore register definition tables extracted from the SDKs: https://svanheule.net/realtek/
> The code archives this info is drawn from, and on which the current implementation is based, can be found here: https://svanheule.net/switches/gpl_source_drops

That's pretty good.

> 
> And that's about it, as far as documentation goes. On the register website, we fill in what we can figure out from the code or hardware behaviour. If you would like to help out with that effort, I can add your Github account to the user whitelist.

I can try to help. This is not my main job focus but I might be able to use some hours on those switches.

> > BTW, is there any progress to get these switches upstream? It looks like it will take a lot of work and cleanup. Tag is done in the ethernet driver while upstream implements dedicated taggers for each protocol.
> 
> There was some initial work on SPI, and IRQ peripherals. I have mainly been working on GPIO (internal and RTL8231). Work on the ethernet part is done entirely downstream by others, with more focus on getting things up and running. I know about some issues, like the tagging, but there will surely be plenty of others.

The DSA upstream is a fast moving target, maybe more than all other related subsystems. It is a complex code but once upstreamed, there will be little work on each kernel upgrade. Il take a look if I can decouple ethernet driver from the switch.

> > And the switch will probably need to use regmap.
> 
> I am aware and started to work on some of the smaller parts of the switchcore. See https://github.com/svanheule/linux/tree/leds/rtl-switch-port which I hope to provide as an RFC to OpenWrt this weekend.

plappermaul commented:
got the serial console on my new DGS-1210-52 F1 up and working. If there is something I can do, just tell me.

luizluca commented:
First, I didn't track if the fixes from @svanheule got merged. They are required for any test. They are not included in this PR.

We still have three issues with this model:

1) tagged vlans go without tags (although my uplink port shouldn't work that way). I'll double check the uplink port once I return to office next week. I also don't know if the issue got fixed meanwhile I'm away.
2) RJ45 combo ports are unstable: only work if initialized by u-boot during a tftp image load. It would be nice to have those ports working independently from u-boot. They are disabled in this PR. I don't know if some device-tree magic can fix that.
3) SFP ports also don't work but the previous issue might be related.

For my setup, tagged vlans is a requirement and I might try to debug that once I return. You can take a look on any of these issues ;-)

plappermaul commented:
With https://github.com/openwrt/openwrt/pull/10319 there comes a lot of fixes and improvements. I will test a mix of these PRs.

svanheule commented:
> We still have three issues with this model:

This is related to the networking code, so not really the area I've been working on until now. If you could create issues for these problems (when you're back at work), maybe someone else will be able to help and there will be more room for in-depth discussion.

flygarn12 commented on target/linux/realtek/image/rtl839x.mk:
```diff
@@ -1,5 +1,13 @@
+  $(Device/d-link_dgs-1210)
+  SOC := rtl8393
+  DEVICE_MODEL := DGS-1210-52
+  DEVICE_PACKAGES += uboot-envtools
```
Have you checked if you need to have uboot-envtools as device package? When I put my -10mp in the realtek config file for uboot-envtools this package became a dependable package and auto included.

flygarn12 reviewed (changes requested):
I have looked at this and leaved some comments.

flygarn12 commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -11,43 +11,31 @@ define Device/allnet_all-sg8208m
-
 define Device/d-link_dgs-1210-10p
   $(Device/d-link_dgs-1210)
+  SOC := rtl8382
```
Ok.

flygarn12 replied:
My comment here doesn't apply any more since I have removed this part from my ongoing PR with the device tree.

Neustradamus commented:
To follow.

luizluca commented:
I might get back to this next week. There are lots of changes related to dgs-1210 that I missed.

luizluca commented:
Superseeded by #10643
